File: Python/algorithms/mathematical/sieve_of_eratosthenes.py
"""
Sieve of Eratosthenes for Generating Prime Numbers

Problem Statement:
Finding all prime numbers up to a given limit is a classic problem in computer science and number theory. 
The naive approach of checking each number for primality individually has a time complexity of O(n√n), 
which becomes inefficient for large limits.

Theory:
The Sieve of Eratosthenes, attributed to the ancient Greek mathematician Eratosthenes, is an efficient 
algorithm to find all primes up to a given number n. 
It works by iteratively marking the multiples of each prime starting from 2 as non-prime. 
Optimizations, such as skipping even numbers and starting marking from i*i, reduce memory usage and runtime.

Algorithm Steps:
1. Initialize a boolean array `sieve` where each index represents a number and False indicates a potential prime.
2. Explicitly include 2 as the only even prime.
3. Iterate through odd numbers starting from 3 up to sqrt(max):
   - If the number is unmarked, it is prime.
   - Mark all odd multiples of this number starting from i*i as non-prime.
4. After marking, collect all unmarked numbers (primes) including 2.

Advantages:
- Time complexity is O(n log log n), efficient for large n.
- Memory usage can be reduced by ignoring even numbers.
- Simple and elegant algorithm with clear implementation.

Use Cases:
- Cryptography
- Number theory computations
- Generating prime numbers for algorithmic problems
"""

def sieve_of_eratosthenes(max_limit):

    if max_limit < 2:
        return []

    # Initialize the sieve with False values
    sieve = [False] * (max_limit + 1)
    # include only 2 -> only even prime number
    primes = [2]

    # iterate -> only odd number, start from three
    for i in range(3, int(pow(max_limit , 0.5)+1),2):
        if not sieve[i]:

            # Mark odd multiples of i starting from i*i
            for j in range(pow(i,2),max_limit+1,2*i):
                sieve[j] = True

    
    # Collect all remaining odd primes
    for k in range(3, max_limit + 1, 2):
        if not sieve[k]:
            primes.append(k)
    
    return primes


# user input 
max_limit = int(input("Enter the limit : "))
prime_numbers = sieve_of_eratosthenes(max_limit)
print(f"Prime numbers are -> {prime_numbers}")

File: CPP/algorithms/mathematical/sieveOfEratosthenes.cpp
/*
Algorithm Name
Sieve of Eratosthenes

Programming Language
C++

Category
Mathematical Algorithms

Difficulty Level
Medium (Intermediate)

Algorithm Description
The Sieve of Eratosthenes is an efficient algorithm to find all prime numbers up to a given limit n. It works by iteratively marking the multiples of each prime number starting from 2.

Steps:

Create a boolean array isPrime[0..n] and initialize all entries as true.

Set isPrime[0] and isPrime[1] to false since 0 and 1 are not prime.

For each number p from 2 to sqrt(n):

If isPrime[p] is true, mark all multiples of p greater than or equal to p*p as false.

All numbers that remain true in the isPrime array are prime numbers.

Time Complexity: O(n log log n) – very efficient for generating primes up to large n.
Space Complexity: O(n) – for the boolean array.

*/
#include <bits/stdc++.h>
using namespace std;

vector<int> sieveOfEratosthenes(int n) {
    vector<bool> isPrime(n + 1, true);
    isPrime[0] = isPrime[1] = false;

    for (int p = 2; p * p <= n; p++) {
        if (isPrime[p]) {
            for (int i = p * p; i <= n; i += p)
                isPrime[i] = false;
        }
    }

    vector<int> primes;
    for (int i = 2; i <= n; i++) {
        if (isPrime[i]) primes.push_back(i);
    }
    return primes;
}

int main() {
    int n;
    cin>>n;
    vector<int> primes = sieveOfEratosthenes(n);
    cout << "Primes up to " << n << ": ";
    for (int p : primes) cout << p << " ";
    cout << endl;
    return 0;
}

/*
Time Complexity: 
𝑂O(nloglogn)

Space Complexity: 
O(n)*/

/*
Input 
100

Output
Primes up to 100: 2 3 5 7 11 13 17 19 23 29 31 37 41 43 47 53 59 61 67 71 73 79 83 89 97 
*/